File: app/src/main/java/cn/edu/tju/notepad/NoteActivity.java
package cn.edu.tju.notepad;

import android.app.Activity;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.net.Uri;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;

import java.io.File;
import java.util.Date;

public class NoteActivity extends AppCompatActivity {
    private EditText editTextTitle, editTextContent;
    private LinearLayout imageContainer;
    private Button btnAddImage;

    private NoteDbHelper noteDbHelper;
    private NoteBean noteBean;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_note);

        // Initialize views
        initViews();

        // Set up data
        setupData();

        // Set up listeners
        setupListeners();
    }

    private void initViews() {
        ImageView imageViewBack = findViewById(R.id.imageViewBack);
        editTextTitle = findViewById(R.id.editTextTitle);
        editTextContent = findViewById(R.id.editTextContent);
        imageContainer = findViewById(R.id.imageContainer);
        btnAddImage = findViewById(R.id.btnAddImage);
    }

    private void setupData() {
        // Get passed data
        Intent intent = getIntent();
        String comeFrom = intent.getStringExtra("ComeFrom");
        noteDbHelper = new NoteDbHelper(NoteActivity.this);

        // If editing an existing note
        if (comeFrom != null && comeFrom.equals("NoteAdapter")) {
            noteBean = (NoteBean) intent.getSerializableExtra("NoteBean");
            if (noteBean != null) {
                editTextTitle.setText(noteBean.getTitle());
                editTextContent.setText(noteBean.getContent());

                // Load images
                loadImages();
            }
        }
    }

    private void setupListeners() {
        // Back button
        findViewById(R.id.imageViewBack).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                finish();
            }
        });

        // Submit button
        Button buttonCommit = findViewById(R.id.buttonCommit);
        buttonCommit.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                saveNote();
            }
        });

        // Add image button
        btnAddImage.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                ImageUtils.pickImageFromGallery(NoteActivity.this);
            }
        });
    }

    private void saveNote() {
        // Get input content
        String title = editTextTitle.getText().toString().trim();
        String content = editTextContent.getText().toString().trim();

        if (title.isEmpty() || content.isEmpty()) {
            Toast.makeText(NoteActivity.this, "标题或内容为空，请补充后再发布", Toast.LENGTH_SHORT).show();
            return;
        }

        Intent intent = getIntent();
        String comeFrom = intent.getStringExtra("ComeFrom");

        if (comeFrom != null && comeFrom.equals("Add")) {
            // Create new note
            String time = String.valueOf(new Date());
            if (noteBean == null) {
                noteBean = new NoteBean(title, content, time);
            } else {
                noteBean.setTitle(title);
                noteBean.setContent(content);
                noteBean.setTime(time);
            }

            // Insert into database
            long result = noteDbHelper.insert(noteBean);
            if (result > 0) {
                Toast.makeText(NoteActivity.this, "发布成功", Toast.LENGTH_SHORT).show();
                finish();
            } else {
                Toast.makeText(NoteActivity.this, "发布失败，请重新发布", Toast.LENGTH_SHORT).show();
            }
        } else if (comeFrom != null && comeFrom.equals("NoteAdapter")) {
            // Edit existing note
            boolean noChanges = title.equals(noteBean.getTitle()) && content.equals(noteBean.getContent());

            if (noChanges) {
                Toast.makeText(NoteActivity.this, "没有修改", Toast.LENGTH_SHORT).show();
                finish();
                return;
            }

            // Show confirmation dialog
            AlertDialog.Builder builder = new AlertDialog.Builder(NoteActivity.this);
            builder.setTitle("是否需要修改？");
            builder.setPositiveButton("确定", new DialogInterface.OnClickListener() {
                @Override
                public void onClick(DialogInterface dialogInterface, int i) {
                    noteBean.setTitle(title);
                    noteBean.setContent(content);
                    noteBean.setTime(String.valueOf(new Date()));

                    // Update database
                    long result = noteDbHelper.update(noteBean);
                    if (result > 0) {
                        Toast.makeText(NoteActivity.this, "修改成功", Toast.LENGTH_SHORT).show();
                        finish();
                    } else {
                        Toast.makeText(NoteActivity.this, "修改失败，请重试！", Toast.LENGTH_SHORT).show();
                    }
                }
            });
            builder.setNegativeButton("取消", null);
            AlertDialog alertDialog = builder.create();
            alertDialog.show();
        }
    }

    private void loadImages() {
        if (noteBean != null && noteBean.getImagePaths() != null) {
            for (String imagePath : noteBean.getImagePaths()) {
                if (!imagePath.isEmpty()) {
                    addImageToContainer(imagePath);
                }
            }
        }
    }

    private void addImageToContainer(String imagePath) {
        try {
            // Create image view
            View imageView = LayoutInflater.from(this).inflate(R.layout.image_item, imageContainer, false);
            ImageView img = imageView.findViewById(R.id.imageViewItem);
            ImageView btnDelete = imageView.findViewById(R.id.btnDeleteImage);

            // Set image
            File imgFile = new File(imagePath);
            if (imgFile.exists()) {
                Bitmap bitmap = BitmapFactory.decodeFile(imagePath);
                img.setImageBitmap(bitmap);

                // Set delete button action
                btnDelete.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        // Remove view from container
                        imageContainer.removeView(imageView);

                        // Remove image path from note
                        if (noteBean != null && noteBean.getImagePaths() != null) {
                            noteBean.getImagePaths().remove(imagePath);
                        }
                    }
                });

                // Add view to container
                imageContainer.addView(imageView);
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (resultCode == Activity.RESULT_OK) {
            if (requestCode == ImageUtils.REQUEST_IMAGE_PICK && data != null) {
                // Gallery selection result
                Uri selectedImage = data.getData();
                if (selectedImage != null) {
                    String imagePath = ImageUtils.copyUriToPrivateStorage(this, selectedImage);

                    if (!imagePath.isEmpty()) {
                        // Add image to UI
                        addImageToContainer(imagePath);

                        // Add image path to note
                        if (noteBean == null) {
                            noteBean = new NoteBean("", "", "");
                        }
                        noteBean.addImagePath(imagePath);
                    }
                }
            }
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);

        if (requestCode == ImageUtils.REQUEST_STORAGE_PERMISSION) {
            // Check if all requested permissions were granted
            boolean allGranted = true;
            for (int result : grantResults) {
                if (result != PackageManager.PERMISSION_GRANTED) {
                    allGranted = false;
                    break;
                }
            }

            if (allGranted) {
                // All permissions granted, can pick image
                ImageUtils.pickImageFromGallery(this);
            } else {
                // User denied some permissions
                Toast.makeText(this, "需要存储权限才能上传图片", Toast.LENGTH_LONG).show();
            }
        }
    }
}

File: app/src/main/java/cn/edu/tju/notepad/NoteActivity.kt
package cn.edu.tju.notepad

import android.content.pm.PackageManager
import android.graphics.BitmapFactory
import android.net.Uri
import android.os.Bundle
import android.widget.Toast
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.activity.enableEdgeToEdge
import androidx.activity.result.contract.ActivityResultContracts
import androidx.compose.foundation.Image
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowBack
import androidx.compose.material.icons.filled.Add
import androidx.compose.material.icons.filled.Check
import androidx.compose.material.icons.filled.Close
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.asImageBitmap
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.lifecycle.lifecycleScope
import kotlinx.coroutines.launch
import cn.edu.tju.notepad.sync.SyncManager
import java.io.File
import java.text.SimpleDateFormat
import java.util.*

class NoteActivity : ComponentActivity() {

    private lateinit var noteDbHelper: NoteDbHelper
    private lateinit var syncManager: SyncManager
    private var noteBean: NoteBean? = null
    private var comeFrom: String = ""

    private val imagePickerLauncher = registerForActivityResult(
        ActivityResultContracts.GetContent()
    ) { uri ->
        uri?.let { handleSelectedImage(it) }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        enableEdgeToEdge()

        noteDbHelper = NoteDbHelper(this)
        syncManager = SyncManager(this, noteDbHelper)
        comeFrom = intent.getStringExtra("ComeFrom") ?: ""
        noteBean = intent.getSerializableExtra("NoteBean", NoteBean::class.java)

        setContent {
            NotepadTheme {
                NoteEditScreen(
                    noteBean = noteBean,
                    comeFrom = comeFrom,
                    onBackPressed = { finish() },
                    onSaveNote = { title, content, imagePaths ->
                        saveNote(title, content, imagePaths)
                    },
                    onAddImage = { addImage() },
                    onDeleteImage = { imagePath ->
                        // 处理图片删除
                        deleteImageFile(imagePath)
                    }
                )
            }
        }
    }

    private fun addImage() {
        if (ImageUtils.checkStoragePermission(this)) {
            imagePickerLauncher.launch("image/*")
        } else {
            ImageUtils.requestStoragePermission(this)
        }
    }

    private fun handleSelectedImage(uri: Uri) {
        try {
            val imagePath = ImageUtils.copyUriToPrivateStorage(this, uri)
            if (imagePath.isNotEmpty()) {
                // 更新当前笔记的图片路径列表
                if (noteBean == null) {
                    noteBean = NoteBean(
                        id = 0,
                        title = "",
                        content = "",
                        time = "",
                        imagePaths = mutableListOf(imagePath),
                        serverId = 0,
                        lastModified = System.currentTimeMillis(),
                        syncStatus = SyncStatus.LOCAL_ONLY,
                        needsUpload = true
                    )
                } else {
                    noteBean?.imagePaths?.add(imagePath)
                    noteBean?.markAsModified()
                }
                Toast.makeText(this, "图片添加成功", Toast.LENGTH_SHORT).show()
            }
        } catch (e: Exception) {
            e.printStackTrace()
            Toast.makeText(this, "图片添加失败: ${e.message}", Toast.LENGTH_SHORT).show()
        }
    }

    private fun saveNote(title: String, content: String, imagePaths: List<String>) {
        if (title.trim().isEmpty() && content.trim().isEmpty()) {
            Toast.makeText(this, "标题和内容不能都为空", Toast.LENGTH_SHORT).show()
            return
        }

        val currentTime = SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.getDefault()).format(Date())

        try {
            when (comeFrom) {
                "Add" -> {
                    val newNote = NoteBean(
                        id = 0,
                        title = title,
                        content = content,
                        time = currentTime,
                        imagePaths = imagePaths.toMutableList(),
                        serverId = 0,
                        lastModified = System.currentTimeMillis(),
                        syncStatus = SyncStatus.LOCAL_ONLY,
                        needsUpload = true
                    )
                    val result = noteDbHelper.insert(newNote)
                    if (result > 0) {
                        Toast.makeText(this, "保存成功", Toast.LENGTH_SHORT).show()

                        // 后台同步（可选）
                        lifecycleScope.launch {
                            try {
                                syncManager.syncNotes()
                            } catch (e: Exception) {
                                // 静默处理同步错误，不影响用户体验
                            }
                        }

                        finish()
                    } else {
                        Toast.makeText(this, "保存失败", Toast.LENGTH_SHORT).show()
                    }
                }
                "NoteAdapter" -> {
                    noteBean?.let { note ->
                        val hasChanges = note.title != title ||
                                note.content != content ||
                                note.imagePaths != imagePaths

                        if (hasChanges) {
                            note.title = title
                            note.content = content
                            note.time = currentTime
                            note.imagePaths = imagePaths.toMutableList()
                            // 标记为已修改，需要同步
                            note.markAsModified()

                            val result = noteDbHelper.update(note)
                            if (result > 0) {
                                Toast.makeText(this, "更新成功", Toast.LENGTH_SHORT).show()

                                // 后台同步（可选）
                                lifecycleScope.launch {
                                    try {
                                        syncManager.syncNotes()
                                    } catch (e: Exception) {
                                        // 静默处理同步错误，不影响用户体验
                                    }
                                }

                                finish()
                            } else {
                                Toast.makeText(this, "更新失败", Toast.LENGTH_SHORT).show()
                            }
                        } else {
                            finish()
                        }
                    }
                }
            }
        } catch (e: Exception) {
            e.printStackTrace()
            Toast.makeText(this, "操作失败: ${e.message}", Toast.LENGTH_SHORT).show()
        }
    }

    private fun deleteImageFile(imagePath: String) {
        try {
            val file = File(imagePath)
            if (file.exists()) {
                file.delete()
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<String>,
        grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        when (requestCode) {
            ImageUtils.REQUEST_STORAGE_PERMISSION -> {
                if (grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                    imagePickerLauncher.launch("image/*")
                } else {
                    Toast.makeText(this, "需要存储权限才能添加图片", Toast.LENGTH_SHORT).show()
                }
            }
        }
    }
}

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun NoteEditScreen(
    noteBean: NoteBean?,
    comeFrom: String,
    onBackPressed: () -> Unit,
    onSaveNote: (String, String, List<String>) -> Unit,
    onAddImage: () -> Unit,
    onDeleteImage: (String) -> Unit
) {
    var title by remember { mutableStateOf(noteBean?.title ?: "") }
    var content by remember { mutableStateOf(noteBean?.content ?: "") }
    var imagePaths by remember { mutableStateOf(noteBean?.imagePaths?.toMutableList() ?: mutableListOf<String>()) }
    var showConfirmDialog by remember { mutableStateOf(false) }

    // 监听noteBean的变化，更新imagePaths
    LaunchedEffect(noteBean?.imagePaths) {
        noteBean?.imagePaths?.let { paths ->
            imagePaths = paths.toMutableList()
        }
    }

    Scaffold(
        modifier = Modifier.fillMaxSize(),
        topBar = {
            TopAppBar(
                title = {
                    Text(
                        text = if (comeFrom == "Add") "新建笔记" else "编辑笔记",
                        fontWeight = FontWeight.SemiBold
                    )
                },
                navigationIcon = {
                    IconButton(onClick = onBackPressed) {
                        Icon(Icons.AutoMirrored.Filled.ArrowBack, contentDescription = "返回")
                    }
                },
                actions = {
                    IconButton(
                        onClick = {
                            if (comeFrom == "NoteAdapter" && noteBean != null) {
                                val hasChanges = noteBean.title != title ||
                                        noteBean.content != content ||
                                        noteBean.imagePaths != imagePaths
                                if (hasChanges) {
                                    showConfirmDialog = true
                                } else {
                                    onBackPressed()
                                }
                            } else {
                                onSaveNote(title, content, imagePaths.toList())
                            }
                        }
                    ) {
                        Icon(Icons.Default.Check, contentDescription = "保存")
                    }
                }
            )
        },
        floatingActionButton = {
            FloatingActionButton(
                onClick = onAddImage,
                containerColor = MaterialTheme.colorScheme.secondary
            ) {
                Icon(Icons.Default.Add, contentDescription = "添加图片")
            }
        }
    ) { paddingValues ->
        Column(
            modifier = Modifier
                .fillMaxSize()
                .padding(paddingValues)
                .padding(16.dp)
                .verticalScroll(rememberScrollState())
        ) {
            // 标题输入框
            OutlinedTextField(
                value = title,
                onValueChange = { title = it },
                label = { Text("标题") },
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(bottom = 16.dp),
                singleLine = true
            )

            // 内容输入框
            OutlinedTextField(
                value = content,
                onValueChange = { content = it },
                label = { Text("内容") },
                modifier = Modifier
                    .fillMaxWidth()
                    .height(300.dp)
                    .padding(bottom = 16.dp),
                maxLines = Int.MAX_VALUE
            )

            // 图片列表
            if (imagePaths.isNotEmpty()) {
                Text(
                    text = "图片 (${imagePaths.size})",
                    style = MaterialTheme.typography.titleMedium,
                    modifier = Modifier.padding(bottom = 8.dp)
                )

                LazyColumn(
                    modifier = Modifier.height(200.dp),
                    verticalArrangement = Arrangement.spacedBy(8.dp)
                ) {
                    items(imagePaths.size) { index ->
                        val imagePath = imagePaths[index]
                        ImageItem(
                            imagePath = imagePath,
                            onDeleteImage = { pathToDelete ->
                                imagePaths.remove(pathToDelete)
                                onDeleteImage(pathToDelete)
                            }
                        )
                    }
                }
            }
        }
    }

    // 确认对话框
    if (showConfirmDialog) {
        AlertDialog(
            onDismissRequest = { showConfirmDialog = false },
            title = { Text("保存修改") },
            text = { Text("是否需要保存修改？") },
            confirmButton = {
                TextButton(
                    onClick = {
                        showConfirmDialog = false
                        onSaveNote(title, content, imagePaths.toList())
                    }
                ) {
                    Text("保存")
                }
            },
            dismissButton = {
                TextButton(
                    onClick = {
                        showConfirmDialog = false
                        onBackPressed()
                    }
                ) {
                    Text("不保存")
                }
            }
        )
    }
}

@Composable
fun ImageItem(
    imagePath: String,
    onDeleteImage: (String) -> Unit
) {
    Box(
        modifier = Modifier.size(120.dp)
    ) {
        Card(
            modifier = Modifier.fillMaxSize(),
            elevation = CardDefaults.cardElevation(defaultElevation = 4.dp),
            shape = RoundedCornerShape(8.dp)
        ) {
            val bitmap = remember(imagePath) {
                try {
                    val file = File(imagePath)
                    if (file.exists()) {
                        BitmapFactory.decodeFile(imagePath)
                    } else {
                        null
                    }
                } catch (e: Exception) {
                    null
                }
            }

            bitmap?.let {
                Image(
                    bitmap = it.asImageBitmap(),
                    contentDescription = "图片",
                    modifier = Modifier.fillMaxSize(),
                    contentScale = ContentScale.Crop
                )
            } ?: Box(
                modifier = Modifier
                    .fillMaxSize()
                    .clip(RoundedCornerShape(8.dp)),
                contentAlignment = Alignment.Center
            ) {
                Text(
                    text = "图片加载失败",
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                    style = MaterialTheme.typography.bodySmall
                )
            }
        }

        // 删除按钮
        IconButton(
            onClick = { onDeleteImage(imagePath) },
            modifier = Modifier.align(Alignment.TopEnd)
        ) {
            Card(
                colors = CardDefaults.cardColors(
                    containerColor = MaterialTheme.colorScheme.error
                ),
                shape = CircleShape
            ) {
                Icon(
                    Icons.Default.Close,
                    contentDescription = "删除图片",
                    tint = MaterialTheme.colorScheme.onError,
                    modifier = Modifier.padding(4.dp)
                )
            }
        }
    }
}